src/tracing/esp32_trace/include/matter/tracing/insights_sys_stats.cpp

```diff
@@ -0,0 +1,158 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "insights_sys_stats.h"
+#include <esp_diagnostics_metrics.h>
+#include <esp_err.h>
+#include <esp_log.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <string.h>
+#include <system/SystemStats.h>
+
+namespace chip {
+namespace System {
+namespace Stats {
+
+InsightsSystemMetrics & InsightsSystemMetrics::GetInstance()
+{
+    static InsightsSystemMetrics instance;
+    return instance;
+}
+
+void InsightsSystemMetrics::SamplingHandler(Layer * systemLayer, void * context)
+{
+    auto instance            = static_cast<InsightsSystemMetrics *>(context);
+    count_t * highwatermarks = GetHighWatermarks();
+    for (int i = 0; i < System::Stats::kNumEntries; i++)
+    {
+        esp_err_t err = esp_diag_metrics_add_uint(instance->mLabels[i], static_cast<int>(highwatermarks[i]));
+        if (err != ESP_OK)
+        {
+            ESP_LOGE(kTag, "Failed to add the metric:%s, err:%d", instance->mLabels[i], err);
+        }
+    }
+    DeviceLayer::SystemLayer().StartTimer(instance->GetSamplingInterval(), SamplingHandler, instance);
+}
+
+CHIP_ERROR InsightsSystemMetrics::Unregister()
+{
+    if (!mRegistered)
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    for (int i = 0; i < System::Stats::kNumEntries; i++)
+    {
+        if (mLabels[i] != nullptr)
+        {
+            esp_err_t err = esp_diag_metrics_unregister(mLabels[i]);
+            if (err != ESP_OK)
+            {
+                ESP_LOGE(kTag, "Failed to unregister metric:%s, err:%d", mLabels[i], err);
+            }
+            free(mLabels[i]);
+            mLabels[i] = nullptr;
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+void InsightsSystemMetrics::SetSamplingHandler(intptr_t arg)
+{
+    auto instance = static_cast<InsightsSystemMetrics *>(reinterpret_cast<void *>(arg));
+
+    if (instance->mTimeout == System::Clock::kZero)
+    {
+        DeviceLayer::SystemLayer().CancelTimer(SamplingHandler, instance);
+    }
+    else
+    {
+        DeviceLayer::SystemLayer().CancelTimer(SamplingHandler, instance);
+        CHIP_ERROR err = DeviceLayer::SystemLayer().StartTimer(instance->mTimeout, SamplingHandler, instance);
+        if (err != CHIP_NO_ERROR)
+        {
+            ESP_LOGE(kTag, "Failed to start the new timer %" CHIP_ERROR_FORMAT, err.Format());
+        }
+    }
+}
+
+CHIP_ERROR InsightsSystemMetrics::SetSamplingInterval(chip::System::Clock::Timeout aTimeout)
+{
+    if (!mRegistered)
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    mTimeout = aTimeout;
+    DeviceLayer::PlatformMgr().ScheduleWork(SetSamplingHandler, reinterpret_cast<intptr_t>(this));
+
+    return CHIP_NO_ERROR;
```

Comment:
I think `this` pointer should work here, and we should return the value of `ScheduleWork` rather than `CHIP_NO_ERROR`.
```suggestion
    return DeviceLayer::PlatformMgr().ScheduleWork(SetSamplingHandler, this);
```